ext/bg/js/dictionary-importer.js

```diff
@@ -275,4 +293,76 @@ class DictionaryImporter {
 
         return [termBank, termMetaBank, kanjiBank, kanjiMetaBank, tagBank];
     }
+
+    async _formatDictionaryTermGlossaryObject(data, context, entry) {
+        switch (data.type) {
+            case 'text':
+                return data.text;
+            case 'image':
+                return await this._formatDictionaryTermGlossaryImage(data, context, entry);
+            default:
+                throw new Error(`Unhandled data type: ${data.type}`);
+        }
+    }
+
+    async _formatDictionaryTermGlossaryImage(data, context, entry) {
+        const dictionary = entry.dictionary;
+        const {path, width: preferredWidth, height: preferredHeight, title, description, pixelated} = data;
+        if (context.media.has(path)) {
+            // Already exists
+            return data;
+        }
+
+        let errorSource = entry.expression;
+        if (entry.reading.length > 0) {
+            errorSource += ` (${entry.reading});`;
+        }
+
+        const file = context.archive.file(path);
+        if (file === null) {
+            throw new Error(`Could not find image at path ${JSON.stringify(path)} for ${errorSource}`);
+        }
+
+        const source = await file.async('base64');
+        const mediaType = mediaUtility.getImageMediaTypeFromFileName(path);
+        if (mediaType === null) {
+            throw new Error(`Could not determine media type for image at path ${JSON.stringify(path)} for ${errorSource}`);
+        }
+
+        let image;
+        try {
+            image = await mediaUtility.loadImage(mediaType, source);
+        } catch (e) {
+            throw new Error(`Could not load image at path ${JSON.stringify(path)} for ${errorSource}`);
+        }
+
+        const width = image.naturalWidth;
+        const height = image.naturalHeight;
+
+        // Create image data
+        const mediaData = {
+            dictionary,
+            path,
+            mediaType,
+            width,
+            height,
+            source
```

Comment:
Is there a reason why this is called `source`? Maybe `base64Data` instead.

Reply:
I thought about "data", but it seemed redundant with "mediaData" and also non-specific as to what the data was. "base64" seemed superfluous to the name in the same way that doing something like "nameUTF8" would be. I chose "source" probably due to "source file" or "source code", but I agree "source" is also not fully specific (curse you English). Maybe "content", since it's similar to how it's used in the context of HTTP? (Content-Type, [etc.](https://www.google.com/search?q=http+content)) I suppose there's then a slight disconnect between using "mediaType" instead of "contentType", but [that's a whole separate story](https://stackoverflow.com/questions/9258108/whats-the-difference-between-mediatype-contenttype-and-mimetype).

Reply:
I brought this up because I couldn't immediately see where the image data was saved to the database before tracing the code back to the line with `const source = await file.async('base64');`. For some time I actually thought that everything is done with the filename only but it felt strange to keep a file handle open to the zip that could be moved at any time.

"content" feels more intuitive to me than "source". "source" might work things like the original image path or dictionary name/ID. (But that's my non-native programming biased English speaker opinion. I've also been surprised many times when hearing a video game term used in the real world differently.)

I agree that `base64*` is not optimal, but at least it's something recognized by (web) developers. It's also used `loadImage` here: https://github.com/FooSoft/yomichan/pull/446/files#diff-50b9f3dfe5b96d91853bf0df8be101b3R55

Reply:
Updated several things to improve API clarity. 7faaf4e45737dd06ab3fdf189bd9c1d26ad1349d...a7e7d546c70c604e986d5c772d842c47fb701eda

However, this has made me notice one thing. This system currently loads images on the backend to validate them, but I don't think this will work with service workers. Not a dealbreaker, it just means that some of the validation will either have to be removed or moved into the settings page during import. There may also be other service worker issues with dictionary import due to how long the operation takes, but those are still unknown. #455

Reply:
> Updated several things to improve API clarity. [7faaf4e...a7e7d54](https://github.com/FooSoft/yomichan/compare/7faaf4e45737dd06ab3fdf189bd9c1d26ad1349d...a7e7d546c70c604e986d5c772d842c47fb701eda)

Thanks! I like the usage of base64 in the function name.

> However, this has made me notice one thing. This system currently loads images on the backend to validate them, but I don't think this will work with service workers. Not a dealbreaker, it just means that some of the validation will either have to be removed or moved into the settings page during import. There may also be other service worker issues with dictionary import due to how long the operation takes, but those are still unknown. #455

I hope the move to manifest v3 isn't going to be a total disaster breaking everything like when Firefox moved from XUL to WebExtensions. At least XUL was making the browser development harder. I guess the background page thing was also a hack in lack of better APIs.